From 0d6797fb589aaf8cef3f6088cce740e40b0bbbc6 Mon Sep 17 00:00:00 2001
From: Pavlin Radoslavov <pavlin@google.com>
Date: Wed, 12 Jul 2017 17:33:42 -0700
Subject: [PATCH] Add a missing check for PAN buffer size before copying data

Bug: 63146237
Test: External script

CVE-2017-0782

Change-Id: I3e9c8a767a8a2a80ff56ccb48c56ca0d4b8c3402
Merged-In: I3e9c8a767a8a2a80ff56ccb48c56ca0d4b8c3402
(cherry picked from commit 1d909399cb4259243dac2e531e3ce6ca1afa77e7)
(cherry picked from commit aa486ad8b5ad6eaef732e5fa7f151495c8c3faf2)
(cherry picked from commit a8a6a17fdfc8d930ba4ad18f92cf4453cc1a219e)
(cherry picked from commit d1145e0af3507e37d4bd25f1833e22c5c716f0ac)
---
 bta/pan/bta_pan_act.c | 11 +++++++++++
 include/bt_target.h   |  4 ++++
 2 files changed, 15 insertions(+)

diff --git a/bta/pan/bta_pan_act.c b/bta/pan/bta_pan_act.c
index 0d82223..ef69547 100644
--- a/bta/pan/bta_pan_act.c
+++ b/bta/pan/bta_pan_act.c
@@ -26,6 +26,8 @@
 
 #if defined(PAN_INCLUDED) && (PAN_INCLUDED == TRUE)
 
+#include <cutils/log.h>
+
 #include "bta_api.h"
 #include "bta_sys.h"
 #include "bd.h"
@@ -206,6 +208,15 @@ static void bta_pan_data_buf_ind_cback(UINT16 handle, BD_ADDR src, BD_ADDR dst,
 
     if ( sizeof(tBTA_PAN_DATA_PARAMS) > p_buf->offset )
     {
+        if (sizeof(BT_HDR) + sizeof(tBTA_PAN_DATA_PARAMS) + p_buf->len >
+            PAN_BUF_SIZE) {
+            android_errorWriteLog(0x534e4554, "63146237");
+            APPL_TRACE_ERROR2("%s: received buffer length too large: %d", __func__,
+                             p_buf->len);
+            GKI_freebuf(p_buf);
+            return;
+        }
+
         /* offset smaller than data structure in front of actual data */
         p_new_buf = (BT_HDR *)GKI_getpoolbuf( PAN_POOL_ID );
         if(!p_new_buf)
diff --git a/include/bt_target.h b/include/bt_target.h
index ac9dc88..3e98642 100644
--- a/include/bt_target.h
+++ b/include/bt_target.h
@@ -657,6 +657,10 @@
 #define AVDT_DATA_POOL_SIZE         GKI_BUF3_SIZE
 #endif
 
+#ifndef PAN_BUF_SIZE
+#define PAN_BUF_SIZE                GKI_BUF3_SIZE
+#endif
+
 #ifndef PAN_POOL_ID
 #define PAN_POOL_ID                 GKI_POOL_ID_3
 #endif
-- 
2.7.4

